sdk/src/Services/S3/Custom/Util/AmazonS3Uri.cs

```diff
@@ -34,6 +34,10 @@ public class AmazonS3Uri
     {
         private const string EndpointPattern = @"^(.+\.)?s3[.-]([a-z0-9-]+)\.";
 
+        private static Regex endpointRegexMatch;
+
+        private static Regex EndpointRegexMatch => endpointRegexMatch ?? (endpointRegexMatch = new Regex(EndpointPattern, RegexOptions.Compiled));
```

Comment:
Why are you lazily initializing the `EndpointRegexMatch` property vs eagerly initializing the field?

Reply:
For consumers of the library that aren't using the functionality, using lazy loading avoids incurring the performance cost of the compiled Regex. Here is blog post covering the topic: http://blog.regexhero.net/2012/03/quick-tip-about-using.html

Reply:
Per my other comment about reusing `EndpointRegexMatch` in the Constructor, is there still a perf benefit?  Ie, is there a case where a consumer could use a static method that doesn't end up calling the instance constructor?

Reply:
EndpointRegexMatch is an expression-bodied declared property, and the lazy loading execution is a benefit because the RegexOptions.Compile option complies at runtime. This means that a consumer of the class that doesn't call the public method IsAmazonS3Endpoint, which calls the EndpointRegexMatch when an Amazon URL, wouldn't take the performance hit.